src/composition.jl

```diff
@@ -61,26 +61,26 @@ function Composition(seq::LongAminoAcidSeq)
     return Composition{AminoAcid}(count_array2dict(counts, symbols(AminoAcidAlphabet())))
 end
 
-function Composition(iter::AbstractMerIterator{T}) where {T<:AbstractMer}
-    counts = Dict{T,Int}()
-    if ksize(T) ≤ 8
+function Composition(iter::AbstractMerIterator{Kmer{A,K,N}}) where {A,K,N}
+    counts = Dict{Kmer{A,K,N},Int}()
+    if K ≤ 8
         # This is faster for short k-mers.
-        counts′ = zeros(Int, 4 ^ ksize(T))
+        counts′ = zeros(Int, 4 ^ K)
         for mer in iter
-            @inbounds counts′[encoded_data(mer.fw) + 1] += 1
+            @inbounds counts′[first(mer.fw.data) + 1] += 1
         end
         for x in eachindex(counts′)
             @inbounds c = counts′[x]
             if c > 0
-                counts[T(x - 1)] = c
+                counts[Kmer{A,K,N}((unsigned(x - 1),))] = c
             end
         end
     else
         for mer in iter
             counts[x] = get(counts, mer.fw, 0) + 1
         end
     end
-    return Composition{T}(counts)
+    return Composition{Kmer{A,K,N}}(counts)
```

Comment:
So that is one thing we probably need to talk about. Because the existence of Composition for kmers is essential kmer counting, and well I have a specific optimised (well kinda optimised right now) for that in KmerAnalysis.jl And there's already some functionality overlap, which we'll need to address.

Reply:
There is also a great opportunity here to do things in parallel and then merge the results.

To frame the scenario, on the outside, we have an iterator that moves across the packed symbol encodings or vector of symbols. In the optimal case, we can directly use a slice or view of the data over which we are iterating. In the other case, there is some intermediate step where the slice or view needs to map into another struct.

If we generalise this, we have a producer (the iterator) whose results are moved through a chain of mappings before being consumed. Now in terms of package responsibility and overlap, if we have something that works in a composable way, we only need to decide where the mapping functions and consumers belong. 

I think the issue of overlap that you point to arises from the `SpacedMerIterator` and `EveryMerIterator` iterators taking on too much responsibility. What about reducing the iterators' responsibility such that they return a tuple of the position and a slice or view of the data over which they iterate?

In a sense, this would make them more like Mer enumerator's; however, they would undoubtedly belong in `BioSequences`.

The construction of the current `MerIterResult` could then be achieved from the position and forward-slice outside of the `AbstractMerEnumerator`. With this approach, we could move `MerIterResult` and other consumers to `KmerAnalysis.jl`. I do not think we'd lose any optimisations by doing this; we'd just be moving or spreading the optimisations across different functions.